janusgraph-core/src/main/java/org/janusgraph/diskstorage/util/ChunkedJobDefinition.java

```diff
@@ -0,0 +1,97 @@
+// Copyright 2023 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.diskstorage.util;
+
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Class which is useful for data processing in chunks.
+ * The class is used to dynamically add new chunks of data into `dataChunks` queue and incrementally process
+ * those chunks over time. At the end of incremental data processing the future result object `result` must
+ * be completed either with the final computed result or with an exception in case data processing is failed at
+ * any stage.
+ *
+ * @param <T> single chunk of data
+ * @param <C> context for chunks processing job resume
+ * @param <R> final computed result
+ */
+public class ChunkedJobDefinition<T, C, R> {
+
+    /**
+     * Dynamic holder for data chunks to process.
+     */
+    private final Queue<T> dataChunks = new ConcurrentLinkedQueue<>();
```

Comment:
I don't know why this needs to be a concurrent queue when you already have a lock to protect it.

Reply:
> I don't know why this needs to be a concurrent queue. I thought you already have a lock to protect it.

The problem is that the lock is for consumer only and not for producer. Producer thread adds data here without any locks while Consumer may remove data from this queue at the same time Producer adds new data. This may result in ConcurrentModificationException. Thus, we should somehow protect this queue from such scenarios (either by using `ConcurrentLinkedQueue` or synchronizing any other queue). 
I'm good with using any other queue implementation here, but the only requirement is that it is protected against  `ConcurrentModificationException`.